### PR TITLE
.NET Language passed through context + SharedModel/.NET Invalid Language Warning

### PR DESCRIPTION
**Pull request by carlos-zamora**
Bugfix for #1899.

* **.NET**: Added PropagateLang to pass `lang` throughout AdaptiveCard contents
* **.NET**: `lang` is checked if valid at parsetime. If invalid, warning is thrown (same warning in Javascript)
* **SharedModel**: `language` is checked if valid at parsetime. If invalid, warning is thrown (same warning in Javascript).

**NOTE**: *.NET* validates `lang` and propagates the corrected `lang` at __parsetime__. *SharedModel* only checks if `language` is valid and attaches warning if applicable. The "correct" language is used/changed at __rendertime__.

### CODE REVIEW

**Comment by khouzam** [comment]
Please update master and merge it into your branch to fix the build break.

**Comment by carlos-zamora** [comment]
@khouzam Oops. A few mistakes on my end. I'll let you know when it's actually ready. Sorry!

File: source/dotnet/Library/AdaptiveCards/AdaptiveColumnSet.cs [hunk 1]
```diff
@@ -38,5 +38,13 @@ public class AdaptiveColumnSet : AdaptiveElement
 #endif
         [DefaultValue(null)]
         public AdaptiveAction SelectAction { get; set; }
+
+        public void SetLang(string value)
+        {
+            foreach(AdaptiveColumn column in Columns)
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards/AdaptiveColumnSet.cs]: need a space here

Comment (carlos-zamora) [reply source/dotnet/Library/AdaptiveCards/AdaptiveColumnSet.cs]: So like this?:
```
public void SetLang (string value) 
{ 
    foreach (AdaptiveColumn column in Columns)
```

File: source/dotnet/Library/AdaptiveCards/Util.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    static class Util
+    {
+        // iterates through each element of the body and sets the language on non-null ColumnSets, Containers, and TextBlocks
+        internal static void PropagateLang(string lang, List<AdaptiveElement> body)
+        {
+            foreach (AdaptiveElement element in body)
+            {
+                string type = element.Type;
+
+                if (type.Equals(AdaptiveColumnSet.TypeName))
```
Comment (paulcam206) [inline source/dotnet/Library/AdaptiveCards/Util.cs]: >if (type.Equals(AdaptiveColumnSet.TypeName)) [](start = 16, length = 44)

this feels like a `switch` statement

File: source/shared/cpp/ObjectModel/SharedAdaptiveCard.cpp [hunk 3]
```diff
@@ -105,6 +105,16 @@ std::shared_ptr<ParseResult> AdaptiveCard::Deserialize(
     std::string language = ParseUtil::GetString(json, AdaptiveCardSchemaKey::Language);
     std::string speak = ParseUtil::GetString(json, AdaptiveCardSchemaKey::Speak);
 
+    // check if language is valid
+    try 
+    {
+        std::locale(language.c_str());
+    }
+    catch (...)
```
Comment (paulcam206) [inline source/shared/cpp/ObjectModel/SharedAdaptiveCard.cpp]: >(...) [](start = 10, length = 5)

can we catch a more specific exception here?

**Comment by carlos-zamora** [comment]
# Valid Languages
Valid languages include the following:
* `null`
* empty string (`""`)
* 2 letters
* 3 letters

Any strings with invalid characters result in warning. Any strings composed of letters only but that are invalid do not throw warning. This matches Javascript behavior.

# .NET: Language ~~Propagation~~ Handling
Language propagation got removed from the PR. Now the language is just passed into the context and used by the textblock... waaay easier.

File: source/shared/cpp/ObjectModel/SharedAdaptiveCard.cpp (hunk 3 above)
Comment (carlos-zamora) [reply source/shared/cpp/ObjectModel/SharedAdaptiveCard.cpp]: replaced by `std::runtime_error`. Still super not specific but that's the best I could find.

File: source/dotnet/Library/AdaptiveCards/Util.cs (hunk 2 above)
Comment (carlos-zamora) [reply source/dotnet/Library/AdaptiveCards/Util.cs]: Removed from PR.

**Pull request merged**